### ISSUE TITLE
Include country ISO 3166-2 code tag for state boundaries

### ISSUE THREAD

**Issue by mmc1718** [openmaptiles/openmaptiles#1567]
Hello all,

From zoom 1, state boundaries for all countries are present in the data. For larger countries this makes sense (US, Russia, Australia, China etc) but for smaller countries (e.g. central Europe, Germany) having them visible makes it difficult to immediately see the country boundaries at low zooms. E.g. [https://www.maptiler.com/maps/#style=basic&mode=2d&position=3.1/48.75/14.97](https://www.maptiler.com/maps/#style=basic&mode=2d&position=3.1/48.75/14.97)

![zoom3](https://github.com/openmaptiles/openmaptiles/assets/65325195/5d6c10e0-8304-497a-906b-eb731a5fa8b4)

There is currently no easy way to differentiate the state boundaries within different countries when styling which means either all states are visible or none are.

Comparing with other maps (Mapbox, Google, Bing) all have states turned off until high zooms for central European countries, with larger countries showing states from lower zooms.

I propose including the ISO 3116-2 code which is already present in the OSM elements (e.g. https://www.openstreetmap.org/relation/62504) in the state boundary features so that it is possible to style states individually by country, by inspecting the ISO code string. I have implemented a version of this in a Planetiler fork but since it breaks the OMT schema it doesn't pass all the tests.

(separate point but this also makes it possible to hide the problem that some state boundaries disappear after zoom 4 (e.g. Ireland) when zooming in, since that is the cut off for Natural Earth boundaries and the OSM equivalents aren't included.)

Interested to hear thoughts.

**Comment by TomPohys** [comment]
Hi, It makes sense to include `iso_a2` code for lower `admin_level` boundary linestrings. Could it be possible to reimplement your planetiler code into SQL? 

You can always change the style and make the `admin_level != 2` much less visible than `admin_level = 2`.

**Comment by mmc1718** [comment]
Thanks for the reply, I would be happy to work on this.

The main difficulty with using admin level in the style is that it also includes state boundaries such as the US states which we would prefer to be kept visible at lower zooms.

Until this change is made another workaround would be to specify the state names to be shown in the style using a filter on the name but this is a bit messy.

**Comment by TomPohys** [comment]
On the other hand, you can make `admin_level = 2` much darker, boosting the countries' border. But anyway, borders with the `iso_a2` code will be helpful.

**Comment by ZeLonewolf** [comment]
I'm not a fan of using the ISO codes in the boundaries, and I think it was a mistake and a hack to do it at the `admin_level=2` level in the first place. What's needed to make osm-carto style boundary renderings is the name of each boundary at each level, represented as a polygon.

I've discussed this in further detail here:
https://github.com/openmaptiles/openmaptiles/pull/1604#issuecomment-1868558482

**Comment by mmc1718** [comment]
@ZeLonewolf do you mean including the name of the country each admin level 2 boundary falls in, instead of the ISO code? The main problem here is how to determine the country associated with a state boundary, since different countries have different rendering requirements for states. It would be great to be able to hide or show state boundaries by country from the style. I am not sure I understand how boundary polygons helps to solve this problem in particular?

**Comment by ZeLonewolf** [comment]
> It would be great to be able to hide or show state boundaries by country from the style.

I think this is out of scope for a general-purpose tile schema and would be more appropriate in a variant schema.  However, the broader point that you raise, which is that sub-national boundaries show up at zoom levels which don't make sense, **are** appropriate to resolve in a general-purpose schema.  I think it's extremely challenging to do that in the SQL and will likely break down under a more complete implementation as I've sketched out in the comment I linked above. Both polygons and boundary segments are required in order to have a "complete" boundary implementation as exists in osm-carto, because you want line styles that represent the lowest-level boundary as well as boundary lables for each touching boundary.

So on one hand I'm concerned about further proliferating the ISO code hacks further into the boundary layer.  On the other hand, I don't think the very real problem you've sketched out should be a style-time decision, it should be done tile-side.  If we need country-specific to determine when to show certain categories of intra-national, that seems hard but certainly doable (we do this for Ireland and the UK in the transportation layers).

**Comment by 1ec5** [comment]
> So on one hand I'm concerned about further proliferating the ISO code hacks further into the boundary layer. On the other hand, I don't think the very real problem you've sketched out should be a style-time decision, it should be done tile-side. If we need country-specific logic to determine when to show certain categories of intra-national, that seems hard but certainly doable (we do this for Ireland and the UK in the transportation layers).

Indeed, older versions of the Mapbox Streets source that used OSM boundaries included only some first-level subnational boundaries at certain zoom levels, depending on the country. I don’t know if there was a whitelist by country, or if it was based on the country’s area or the average area of the subnational boundary. The schema didn’t expose this decision explicitly. The Streets source did indicate ISO&nbsp;3166-1&nbsp;alpha-2 codes on boundaries, but only for more selective filtering – for example, if you only want to show Canadian provincial boundaries on a Canada-focused map.

By the way, the alpha-3 codes that OMT uses are very inconvenient to use compared to the alpha-2 codes that most platforms have standardized on – see ZeLonewolf/openstreetmap-americana#753 for the gargantuan workaround I had to put in just to transform alpha-3 codes to names in the user’s preferred language. Also, note that ISO&nbsp;3166-2 still only extends to first-order administrative subdivisions (`admin_level=4`), whereas some countries have very significant second-order subdivisions (e.g., `admin_level=6`) that warrant boundary edge labels at higher zoom levels. If the subnational boundaries can be filtered server-side, then I think ISO&nbsp;3166-2 codes could be useful for some purposes, but not as useful as names.
